index.js

```diff
@@ -77,91 +103,59 @@ var agent;
+  if (traceAgent && !config.forceNewAgent_) { // already started.
     throw new Error('Cannot call start on an already started agent.');
+  } else if (traceAgent) {
+    // For unit tests only.
+    // Undoes initialization that occurred last time start() was called.
+    stop();
   }
 
   if (!config.enabled) {
-    return traceApi;
+    return traceAgent;
   }
 
-  if (config.logLevel < 0) {
-    config.logLevel = 0;
-  } else if (config.logLevel >= common.logger.LEVELS.length) {
-    config.logLevel = common.logger.LEVELS.length - 1;
-  }
   var logger = common.logger({
     level: common.logger.LEVELS[config.logLevel],
     tag: '@google-cloud/trace-agent'
   });
 
-  if (config.projectId) {
-    logger.info('Locally provided ProjectId: ' + config.projectId);
-  }
-
-  if (onUncaughtExceptionValues.indexOf(config.onUncaughtException) === -1) {
-    logger.error('The value of onUncaughtException should be one of ',
-      onUncaughtExceptionValues);
-    throw new Error('Invalid value for onUncaughtException configuration.');
-  }
-
-  var headers = {};
-  headers[constants.TRACE_AGENT_REQUEST_HEADER] = 1;
-
   if (modulesLoadedBeforeTrace.length > 0) {
     logger.error('Tracing might not work as the following modules ' +
       'were loaded before the trace agent was initialized: ' +
       JSON.stringify(modulesLoadedBeforeTrace));
   }
+  // CLS namespace for context propagation
+  cls.createNamespace();
+  TraceWriter.create(logger, config);
 
-  if (typeof config.projectId === 'undefined') {
-    // Queue the work to acquire the projectId (potentially from the
-    // network.)
-    gcpMetadata.project({
-      property: 'project-id',
-      headers: headers
-    }, function(err, response, projectId) {
-      if (response && response.statusCode !== 200) {
-        if (response.statusCode === 503) {
-          err = new Error('Metadata service responded with a 503 status ' +
-            'code. This may be due to a temporary server error; please try ' +
-            'again later.');
-        } else {
-          err = new Error('Metadata service responded with the following ' +
-            'status code: ' + response.statusCode);
-        }
-      }
-      if (err) {
-        logger.error('Unable to acquire the project number from metadata ' +
-          'service. Please provide a valid project number as an env. ' +
-          'variable, or through config.projectId passed to start(). ' + err);
-        if (traceApi.isActive()) {
-          agent.stop();
-          traceApi.disable_();
-          pluginLoader.deactivate();
-        }
-        return;
-      }
-      config.projectId = projectId;
-    });
-  } else if (typeof config.projectId !== 'string') {
+  traceAgent = new TraceAgent('Custom Span API', logger, config);
+  pluginLoader.activate(logger, config);
+
+  // Get metadata.
+  if (typeof config.projectId !== 'string' && typeof config.projectId !== 'undefined') {
```

Comment:
Discussed in person, moving it might result in a semver-minor change, so would do in follow-up PR (this one is meant to be patch-only).